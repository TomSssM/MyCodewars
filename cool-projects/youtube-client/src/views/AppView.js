export default class AppView {
  constructor(data) {
    this.slider = data.slider;
    this.btnContainer = data.btnContainer;
    this.search = data.search;
    this.loadIndicator = data.loadIndicator;
    this.template = data.template;
    this.defaultCardWidth = data.defaultCardWidth;
    this.prevActiveBtn = null;
    this.btnHighlighted = false;
    this.toolTipElem = AppView.renderToolTip();
    document.body.append(this.toolTipElem);
  }

  renderControlBtn(html) {
    const btn = document.createElement('button');
    btn.classList.add('js-control-btn');
    btn.classList.add('scroll-controls__button');
    btn.innerHTML = html;
    this.btnContainer.append(btn);
  }

  renderLoading() {
    this.loadIndicator.hidden = false;
    this.loadIndicator.classList.add('scroll__load-indicator--loading');
  }

  removeLoading() {
    this.loadIndicator.classList.remove('scroll__load-indicator--loading');
    this.loadIndicator.hidden = true;
  }

  renderErrorMsg(msg) {
    if (this.errorElem) return;
    this.errorElem = document.createElement('div');
    this.errorElem.classList.add('error-box');
    this.errorElem.innerHTML = msg;
    document.body.append(this.errorElem);
  }

  removeErrorMsg() {
    this.errorElem.remove();
  }

  renderCards(data) {
    if (!data.items) return;
    if (localStorage.getItem('previousVideos') === null) {
      localStorage.setItem('previousVideos', JSON.stringify(data));
    }
    this.slider.style.display = 'grid';
    let title;
    let url;
    let imgUrl;
    let description;
    let channel;
    let viewCount;
    let date;
    let card;
    let cardHTML;
    data.items.forEach((item) => {
      ({ title } = item.snippet);
      title = AppView.restrictSize(title, 30);

      url = `https://www.youtube.com/watch?v=${item.id}`;
      imgUrl = item.snippet.thumbnails.high.url;

      ({ description } = item.snippet);
      description = AppView.restrictSize(description, 100);

      channel = item.snippet.channelTitle;
      channel = AppView.restrictSize(channel, 12);

      ({ viewCount } = item.statistics);
      date = item.snippet.publishedAt;
      date = date.slice(0, 10);

      card = document.createElement('div');
      cardHTML = this.createCardHTML({
        title,
        url,
        imgUrl,
        description,
        channel,
        viewCount,
        date,
      });
      card.classList.add('scroll-card');
      card.innerHTML = cardHTML;
      this.slider.append(card);
    });
  }

  createCardHTML(data) {
    return this.template(data);
  }

  resizeCards(cardWidth) {
    this.slider.style.gridAutoColumns = `${cardWidth}px`;
  }

  static restrictSize(input, size) {
    if (input.length < size) return input;
    const res = input;
    return `${res.slice(0, size)}...`;
  }

  offetSlider(val) {
    this.slider.style.transform = `translateX(${val}px)`;
  }

  reloadSlider() {
    this.slider.innerHTML = '';
    this.slider.style.transform = '';
    this.slider.style.display = '';
  }

  highlightBtn(pageNum) {
    this.btnHighlighted = true;
    const prevActiveBtn = this.btnContainer.children[this.prevActiveBtn];
    if (prevActiveBtn) {
      prevActiveBtn.classList.remove('scroll-controls__button--active');
    }
    this.btnContainer.children[pageNum].classList.add('scroll-controls__button--active');
    this.prevActiveBtn = pageNum;
  }

  offsetBtns(val) {
    for (let i = 0; i < this.btnContainer.children.length; i += 1) {
      const btn = this.btnContainer.children[i];
      btn.innerHTML = Number(btn.innerHTML) + val;
    }
  }

  get selectedBtn() {
    return this.btnContainer.children[this.prevActiveBtn];
  }

  makeBtnsFromTemplate(template) {
    if (this.btnContainer.children.length > template.length) {
      while (this.btnContainer.children.length > template.length) {
        this.btnContainer.lastElementChild.remove();
      }
    } else if (this.btnContainer.children.length < template.length) {
      while (this.btnContainer.children.length < template.length) {
        this.renderControlBtn();
      }
    }

    template.forEach((btnTemp, i) => {
      this.btnContainer.children[i].innerHTML = btnTemp.innerHTML;
      if (btnTemp.selected) {
        this.highlightBtn(i);
      }
    });
  }

  getBtnWith(innerHTML = '') {
    const btnNumber = String(innerHTML);
    for (let i = 0; i < this.btnContainer.children.length; i += 1) {
      const btn = this.btnContainer.children[i];
      if (btn.innerHTML === btnNumber) return btn;
    }
    return null;
  }

  deleteBtnsUntil(quantity) {
    while (this.btnContainer.children.length > quantity) {
      this.btnContainer.lastElementChild.remove();
    }
  }

  showToolTip(btn) {
    this.showingToolTipFor = btn;
    this.toolTipElem.innerHTML = btn.innerHTML;
    this.positionTooltipAt(btn);
    setTimeout(() => {
      if (this.showingToolTipFor !== btn) return;
      this.toolTipElem.classList.add('tooltip--fadein');
    }, 500);
  }

  clearToolTip() {
    if (!this.showingToolTipFor) return;
    this.showingToolTipFor = null;
    if (!this.toolTipElem.classList.contains('tooltip--fadein')) return;
    const cb = () => {
      this.toolTipElem.innerHTML = '';
      this.toolTipElem.removeEventListener('transitionend', cb);
    };
    this.toolTipElem.classList.remove('tooltip--fadein');
    this.toolTipElem.addEventListener('transitionend', cb);
  }

  static renderToolTip() {
    const tooltip = document.createElement('div');
    tooltip.classList.add('tooltip');
    return tooltip;
  }

  reset() {
    this.reloadSlider();
    this.btnContainer.innerHTML = '';
    this.prevActiveBtn = null;
    this.btnHighlighted = false;
    this.clearToolTip();
  }

  positionTooltipAt(btn) {
    const coords = btn.getBoundingClientRect();
    const top = coords.top - this.toolTipElem.offsetHeight - 10;
    const left = coords.left + btn.offsetWidth / 2 - this.toolTipElem.offsetWidth / 2;
    this.toolTipElem.style.top = `${top}px`;
    this.toolTipElem.style.left = `${left}px`;
  }
}
