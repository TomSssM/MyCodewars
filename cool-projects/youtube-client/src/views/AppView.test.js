import _ from 'lodash';
import AppView from './AppView';

document.body.innerHTML = (
  `<main class="main">
    <input id="js-search" type="text" class="scroll-search">
    <div class="scroll">
      <div id="js-loading" class="scroll__load-indicator"></div>
      <div id="js-slider" class="scroll__slider"></div>
    </div>
    <div id="js-btn-container" class="scroll-controls"></div>
  </main>`
);

const view = new AppView({
  slider: document.querySelector('#js-slider'),
  btnContainer: document.querySelector('#js-btn-container'),
  search: document.querySelector('#js-search'),
  loadIndicator: document.querySelector('#js-loading'),
  defaultCardWidth: 300,
  template: _.template(
    `<div class="scroll-card__link-container">
          <img src="<%-imgUrl%>" class="scroll-card__link-image">
          <a href="<%-url%>" class="scroll-card__link" class="scroll-card__link-text"><%-title%></a>
        </div>
        <div class="scroll-card__row">
          <i class="fas fa-male scroll-card__icon"></i>
          <span class="scroll-card__text"><%-channel%></span>
        </div>
        <div class="scroll-card__row">
          <i class="fas fa-eye scroll-card__icon"></i>
          <span class="scroll-card__text"><%-viewCount%></span>
        </div>
        <div class="scroll-card__row">
          <i class="fas fa-calendar-plus scroll-card__icon"></i>
          <span class="scroll-card__text"><%-date%></span>
        </div>
        <p class="scroll-card__description"><%-description%></p>`,
  ),
});

describe('AppView.prototype.renderCards', () => {
  test('renders the cards correctly based on fetched data', () => {
    const data = {
      items: [
        {
          id: 'videoID123',
          snippet: {
            title: 'our-title--our-title--our-title--our-title--our-title',
            thumbnails: {
              high: {
                url: 'image-url',
              },
            },
            description: 'lotsofdecsdfsdf'.repeat(120),
            channelTitle: 'lotsofdecsdfsdflotsofdecsdfsdf',
            publishedAt: 'dd-mm-yyyyfsdfsdfdds',
          },
          statistics: {
            viewCount: 120,
          },
        },
        {
          id: 'videoID123',
          snippet: {
            title: 'our-title--our-title--our-title--our-title--our-title',
            thumbnails: {
              high: {
                url: 'image-url',
              },
            },
            description: 'lotsofdecsdfsdf'.repeat(120),
            channelTitle: 'lotsofdecsdfsdflotsofdecsdfsdf',
            publishedAt: 'dd-mm-yyyyfsdfsdfdds',
          },
          statistics: {
            viewCount: 120,
          },
        },
      ],
    };
    view.renderCards(data);
    expect(document.body.innerHTML).toMatchSnapshot();
  });
});
