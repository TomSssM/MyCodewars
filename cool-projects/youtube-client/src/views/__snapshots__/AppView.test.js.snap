// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppView.prototype.renderCards renders the cards correctly based on fetched data 1`] = `
"<main class=\\"main\\">
    <input id=\\"js-search\\" type=\\"text\\" class=\\"scroll-search\\">
    <div class=\\"scroll\\">
      <div id=\\"js-loading\\" class=\\"scroll__load-indicator\\"></div>
      <div id=\\"js-slider\\" class=\\"scroll__slider\\" style=\\"display: grid;\\"><div class=\\"scroll-card\\"><div class=\\"scroll-card__link-container\\">
          <img src=\\"image-url\\" class=\\"scroll-card__link-image\\">
          <a href=\\"https://www.youtube.com/watch?v=videoID123\\" class=\\"scroll-card__link\\">our-title--our-title--our-titl...</a>
        </div>
        <div class=\\"scroll-card__row\\">
          <i class=\\"fas fa-male scroll-card__icon\\"></i>
          <span class=\\"scroll-card__text\\">lotsofdecsdf...</span>
        </div>
        <div class=\\"scroll-card__row\\">
          <i class=\\"fas fa-eye scroll-card__icon\\"></i>
          <span class=\\"scroll-card__text\\">120</span>
        </div>
        <div class=\\"scroll-card__row\\">
          <i class=\\"fas fa-calendar-plus scroll-card__icon\\"></i>
          <span class=\\"scroll-card__text\\">dd-mm-yyyy</span>
        </div>
        <p class=\\"scroll-card__description\\">lotsofdecsdfsdflotsofdecsdfsdflotsofdecsdfsdflotsofdecsdfsdflotsofdecsdfsdflotsofdecsdfsdflotsofdecs...</p></div><div class=\\"scroll-card\\"><div class=\\"scroll-card__link-container\\">
          <img src=\\"image-url\\" class=\\"scroll-card__link-image\\">
          <a href=\\"https://www.youtube.com/watch?v=videoID123\\" class=\\"scroll-card__link\\">our-title--our-title--our-titl...</a>
        </div>
        <div class=\\"scroll-card__row\\">
          <i class=\\"fas fa-male scroll-card__icon\\"></i>
          <span class=\\"scroll-card__text\\">lotsofdecsdf...</span>
        </div>
        <div class=\\"scroll-card__row\\">
          <i class=\\"fas fa-eye scroll-card__icon\\"></i>
          <span class=\\"scroll-card__text\\">120</span>
        </div>
        <div class=\\"scroll-card__row\\">
          <i class=\\"fas fa-calendar-plus scroll-card__icon\\"></i>
          <span class=\\"scroll-card__text\\">dd-mm-yyyy</span>
        </div>
        <p class=\\"scroll-card__description\\">lotsofdecsdfsdflotsofdecsdfsdflotsofdecsdfsdflotsofdecsdfsdflotsofdecsdfsdflotsofdecsdfsdflotsofdecs...</p></div></div>
    </div>
    <div id=\\"js-btn-container\\" class=\\"scroll-controls\\"></div>
  </main><div class=\\"tooltip\\"></div>"
`;
